firefox-ios/Client/Frontend/Browser/BrowserPrompts.swift

```diff
@@ -17,36 +18,52 @@ protocol JSPromptAlertControllerDelegate: AnyObject {
 /// a runtime exception is thrown.
 class JSPromptAlertController: UIAlertController {
     var alertInfo: JSAlertInfo?
+    var logger: Logger = DefaultLogger.shared
+    /// The completion handler that is expected to be called by the WebView delegate.
+    /// It will be called in the de initializer if it is not nil.
+    /// If the completion Handler is nil it means it was already called by the alert actions
+    var completionHandler: VoidReturnCallback?
 
     weak var delegate: JSPromptAlertControllerDelegate?
 
     override func viewDidDisappear(_ animated: Bool) {
         super.viewDidDisappear(animated)
         delegate?.promptAlertControllerDidDismiss(self)
     }
+
+    deinit {
+        if let completionHandler {
+            logger.log(
+                "JS Prompt Alert contrller was deinited with a non nil completionHandler",
+                level: .fatal,
+                category: .webview
+            )
+            completionHandler()
+        }
+    }
 }
 
-/**
- *  An JSAlertInfo is used to store information about an alert we want to show either immediately or later.
- *  Since alerts are generated by web pages and have no upper limit it would be unwise to allocate a
- *  UIAlertController instance for each generated prompt which could potentially be queued in the background.
- *  Instead, the JSAlertInfo structure retains the relevant data needed for the prompt along with a copy
- *  of the provided completionHandler to let us generate the UIAlertController when needed.
- */
+/// An JSAlertInfo is used to store information about an alert we want to show either immediately or later.
+/// Since alerts are generated by web pages and have no upper limit it would be unwise to allocate a
+/// UIAlertController instance for each generated prompt which could potentially be queued in the background.
+/// Instead, the JSAlertInfo structure retains the relevant data needed for the prompt along with a copy
+/// of the provided completionHandler to let us generate the UIAlertController when needed.
 protocol JSAlertInfo {
     func alertController() -> JSPromptAlertController
 }
 
 struct MessageAlert: JSAlertInfo {
     let message: String
     let frame: WKFrameInfo
+    let completionHandler: VoidReturnCallback
```

Comment:
This is the same fix added in https://github.com/mozilla-mobile/firefox-ios/pull/22392/.

The difficulty is that this actually exacerbated the crash rate once it was released. So unfortunately we can't just roll this back in as-is (it was reverted because it actually decreased stability). This is called out in the parent ticket, but we need to identify the crash scenarios that increased in production with this fix in place. Were you able to do any digging into that?

Reply:
Yes true now i moved in the defer of the webView delegate so it gets always called. I don't if that is a problem that we call the completionHandler and then possibly present the Alert later. i think it is not a big deal for `MessageAlert` since the completionHandler doesn't require passing parameters